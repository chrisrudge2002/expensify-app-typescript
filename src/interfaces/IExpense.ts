export default interface IExpense {
    amount: number;
    createdAt: number;
    description: string;
    id?: string;
    note: string;
}
