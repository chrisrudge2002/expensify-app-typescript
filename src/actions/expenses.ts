import IExpense from "../interfaces/IExpense";
import { expenseActionTypes } from "./types";

export const startAddExpense = ({ description = "", note = "", amount = 0, createdAt = 0 }: IExpense) => {
    return {
        expense: {
            amount,
            createdAt,
            description,
            note,
        },
        type: expenseActionTypes.START_ADD_EXPENSE,
    };
};

export const startEditExpense = (id: string, updates = {}) => {
    return {
        id,
        type: expenseActionTypes.START_EDIT_EXPENSE,
        updates,
    };
};

export const startRemoveExpense = (id: string = "") => {
    return {
        id,
        type: expenseActionTypes.START_REMOVE_EXPENSE,
    };
};

export const addExpense = (expense: IExpense) => {
    return {
        expense,
        type: expenseActionTypes.ADD_EXPENSE,
    };
};

export const getExpenses = () => {
    return {
        type: expenseActionTypes.GET_EXPENSES,
    };
};

export const gotExpenses = (expenseList: IExpense[]) => {
    return {
        expenseList,
        type: expenseActionTypes.GOT_EXPENSES,
    };
};

export const editExpense = (id: string, updates = {}) => {
    return {
        id,
        type: expenseActionTypes.EDIT_EXPENSE,
        updates,
    };
};

export const removeExpense = (id: string) => {
    return {
        id,
        type: expenseActionTypes.REMOVE_EXPENSE,
    };
};
