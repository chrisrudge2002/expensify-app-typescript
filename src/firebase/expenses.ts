import * as firebase from "firebase";
import IExpense from "../interfaces/IExpense";
import database from "./firebase";

export const addExpenseToDatabase = async (expense: IExpense, userid: string) => {
    const result = await database.ref(`users/${userid}/expenses`).push(expense);
    return {
        ...expense,
        id: result.key,
    };
};

export const editExpenseInDatabase = async (id: string, updates: {}, userid: string) => {
    await database.ref(`users/${userid}/expenses/${id}`).update({...updates});
};

export const removeExpenseFromDatabase = async (id: string, userid: string) => {
    await database.ref(`users/${userid}/expenses/${id}`).remove();
};

export const getExpensesFromDatabase = async (userid: string) => {
    const snapshot = await database.ref(`users/${userid}/expenses`).once("value");
    const expenses: IExpense[] = [];
    snapshot.forEach((item: firebase.database.DataSnapshot) => {
        expenses.push({
            ...item.val(),
            id: item.key,
        });
    });
    return expenses;
};
