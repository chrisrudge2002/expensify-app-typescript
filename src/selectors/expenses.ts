import * as moment from "moment";
import IExpense from "../interfaces/IExpense";
import IFilterState from "../interfaces/IFilterState";

export default (expenses: IExpense[], { text, sortBy, startDate, endDate }: IFilterState) => {
    return expenses.filter((e: IExpense) => {
        const createdAtMoment = moment(e.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;
        const textMatch = (!text || e.description.toLowerCase().includes(text.toLowerCase()));

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a: IExpense, b: IExpense) => {
        if (sortBy === "amount") {
            return (a.amount < b.amount ? 1 : -1);
        }

        return (a.createdAt < b.createdAt ? 1 : -1);
    });
};
