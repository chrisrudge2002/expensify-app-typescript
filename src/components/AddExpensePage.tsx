import * as React from "react";
import { connect } from "react-redux";
import { RouterProps } from "react-router";
import { AnyAction, Dispatch } from "redux";
import * as expenseActions from "../actions/expenses";
import IExpense from "../interfaces/IExpense";
import ExpenseForm from "./ExpenseForm";

interface IAddExpensePageProps extends RouterProps {
    startAddExpense: (expense: IExpense) => AnyAction;
}

export class AddExpensePage extends React.Component<IAddExpensePageProps> {
    public render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm onFormSubmit={this.onSubmit} />
                </div>
            </div>
        );
    }
    private onSubmit = (expense: IExpense) => {
        this.props.startAddExpense(expense);
        this.props.history.push("/");
    }
}

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
    startAddExpense: (expense: IExpense) => (dispatch(expenseActions.startAddExpense(expense)) as AnyAction),
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
