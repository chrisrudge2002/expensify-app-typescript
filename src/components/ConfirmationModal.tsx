import * as React from "react";
import { connect } from "react-redux";
import { AnyAction, Dispatch } from "redux";
import * as Modal from "react-modal";
import * as expenseActions from "../actions/expenses";

interface IConfirmationModalProps {
    description: string;
    id: string | undefined;
    isOpen: boolean;
    history: any;
    onCancelRemoval: () => void;
    startRemoveExpense: (id: string) => AnyAction;
}

export class ConfirmationModal extends React.Component<IConfirmationModalProps> {
    public render() {
        return (
            <Modal
                isOpen={!!this.props.isOpen}
                onRequestClose={this.props.onCancelRemoval}
                contentLabel="Confirm Removal"
                closeTimeoutMS={200}
                className="modal"
            >
            <h3 className="modal__title">Remove Expense</h3>
            <p className="modal__body">{`Are you sure you want to remove the expense '${this.props.description}'?`}</p>
            <button className="button button--confirm-yes" onClick={this.onConfirmRemoval}>Yes</button>
            <button className="button button--confirm-no" onClick={this.props.onCancelRemoval}>No</button>
          </Modal>
        );
    }
    private onConfirmRemoval = () => {
        if (this.props.id && this.props.startRemoveExpense) {
            this.props.startRemoveExpense(this.props.id);
            this.props.history.push("/");
        }
    }
}

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
    startRemoveExpense: (id: string) => (dispatch(expenseActions.startRemoveExpense(id)) as AnyAction),
});

export default connect(undefined, mapDispatchToProps)(ConfirmationModal);