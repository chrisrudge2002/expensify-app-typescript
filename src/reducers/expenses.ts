import { AnyAction } from "redux";
import { authActionTypes, expenseActionTypes } from "../actions/types";
import IExpense from "../interfaces/IExpense";

const defaultState: IExpense[] = [];

export default (state = defaultState, action: AnyAction) => {
    switch (action.type) {
        case expenseActionTypes.ADD_EXPENSE:
            return [
                ...state,
                action.expense,
            ];
        case expenseActionTypes.EDIT_EXPENSE:
            return state.map((e) => {
                if (e.id === action.id) {
                    return {
                        ...e,
                        ...action.updates,
                    };
                }
                return e;
            });
        case expenseActionTypes.GOT_EXPENSES:
            return action.expenseList;
        case authActionTypes.LOGGED_OUT:
            return [];
        case expenseActionTypes.REMOVE_EXPENSE:
            return state.filter((e) => e.id !== action.id);
        default:
            return state;
    }
};
