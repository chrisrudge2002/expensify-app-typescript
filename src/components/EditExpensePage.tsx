import * as React from "react";
import { connect } from "react-redux";
import { RouteComponentProps } from "react-router";
import { AnyAction, Dispatch } from "redux";
import * as expenseActions from "../actions/expenses";
import IAppState from "../interfaces/IAppState";
import IExpense from "../interfaces/IExpense";
import ExpenseForm from "./ExpenseForm";
import ConfirmationModal from "./ConfirmationModal";

interface IEditExpensePageProps extends RouteComponentProps<any> {
    expense: IExpense;
    startEditExpense: (id: string, updates: {} ) => AnyAction;
}

interface IEditExpensePageState {
    confirmModalIsOpen: boolean;
}

export class EditExpensePage extends React.Component<IEditExpensePageProps, IEditExpensePageState> {
    constructor(props: IEditExpensePageProps) {
        super(props);

        this.state = {
            confirmModalIsOpen: false,
        };
    }
    public render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm expense={this.props.expense} onFormSubmit={this.onSubmit} />
                    <button
                        className="button button--secondary"
                        onClick={this.onRemove}>
                            Remove Expense
                    </button>
                </div>
                <ConfirmationModal 
                    id={this.props.expense.id}
                    description={this.props.expense.description}
                    isOpen={this.state.confirmModalIsOpen}
                    onCancelRemoval={this.onCancelRemoval}
                    history={this.props.history} />
            </div>
        );
    }
    private onCancelRemoval = () => {
        this.setState(() => ({ confirmModalIsOpen: false }));
    }
    private onSubmit = (expense: IExpense) => {
        if (this.props.expense.id) {
            this.props.startEditExpense(this.props.expense.id, expense);
            this.props.history.push("/");
        }
    }
    private onRemove = () => {
        if (this.props.expense.id) {
            this.setState(() => ({ confirmModalIsOpen: true }));
        }
    }
}

const mapStateToProps = (state: IAppState, props: IEditExpensePageProps) => ({
    expense: state.expenses.find((e: IExpense) => e.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
    startEditExpense: (id: string, expense: IExpense) =>
        (dispatch(expenseActions.startEditExpense(id, expense)) as AnyAction),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
