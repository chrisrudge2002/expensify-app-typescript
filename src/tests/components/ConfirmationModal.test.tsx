import { shallow, ShallowWrapper } from "enzyme";
import * as React from "react";
import { ConfirmationModal } from "../../components/ConfirmationModal";
import IExpense from "../../interfaces/IExpense";
import expenses from "../fixtures/expenses";

let onCancelRemoval: jest.Mock; let startRemoveExpense: jest.Mock;
let history: { push: jest.Mock };
let wrapper: ShallowWrapper<React.Component>;

beforeEach(() => {
    history = { push: jest.fn() };
    onCancelRemoval = jest.fn();
    startRemoveExpense = jest.fn();
    wrapper = shallow(<ConfirmationModal
        id={(expenses as IExpense[])[1].id}
        description={(expenses as IExpense[])[1].description}
        isOpen={true}
        onCancelRemoval={onCancelRemoval}
        startRemoveExpense={startRemoveExpense}
        history={history}
    />);
});

test("should render ConfirmationModal correctly", () => {
    expect(wrapper).toMatchSnapshot();
});

test("should handle onCancelRemoval (cancel remove expense)", () => {
    wrapper.find("button").at(1).simulate("click");
    expect(onCancelRemoval).toHaveBeenCalled();
});

test("should handle onRemove (remove expense)", () => {
    wrapper.find("button").at(0).simulate("click");
    expect(history.push).toHaveBeenCalledWith("/");
    expect(startRemoveExpense).toHaveBeenLastCalledWith(expenses[1].id);
});
