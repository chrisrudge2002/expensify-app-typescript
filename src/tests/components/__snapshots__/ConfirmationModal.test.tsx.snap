// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ConfirmationModal correctly 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="Confirm Removal"
  isOpen={true}
  onRequestClose={[MockFunction]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <h3
    className="modal__title"
  >
    Remove Expense
  </h3>
  <p
    className="modal__body"
  >
    Are you sure you want to remove the expense 'Rent'?
  </p>
  <button
    className="button button--confirm-yes"
    onClick={[Function]}
  >
    Yes
  </button>
  <button
    className="button button--confirm-no"
    onClick={[MockFunction]}
  >
    No
  </button>
</Modal>
`;
